Adapter Updates

Skorpeo opened the pull request:
Modifications to allow use of plugin with adapter version 7.3.0

Skorpeo commented:
Hi, I am not a javascript guy, not a ios guy, not a cordova guy :/.  I am also not familiar with pull requests or how to fix above...but it looks like the errors  are for android and due to adapter.js included.  These fixes are only for ios...

hthetiot commented:
@Skorpeo Travis is not happy with adapter.js lint on all tested platform (ios, android, web) :)

<img width="975" alt="Screen Shot 2019-09-10 at 8 30 54 PM" src="https://user-images.githubusercontent.com/8635/64640286-ea2b2100-d409-11e9-912f-97e1c3deec44.png">
Source: https://travis-ci.org/cordova-rtc/cordova-plugin-iosrtc/jobs/582971695#L266

We do not distribute adapter.js could you remove this files from this PR that will make travis and me happy :) Finally can you after removing extras/adaper*.js files, can you run `npm run build` and commit dist generated files.

Just to confirm your fix does work with original adapter.js right since you fixed YaetiEventTarget usage or prototype extensions ? Please confirm.

I will make a full test before merging.

Thank you again for your contribution.

hthetiot commented:
> Hi, I am not a javascript guy, not a ios guy, not a cordova guy :/. 
you are now, one of the best PR i have seen on the JS side so fare.

Big congrats.

hthetiot commented:
Note: iosrtc does not include, will not include and should not require adapter.js but will support it when and if included into the application.

> Question: But adapter should be included in the webrtc ios cordova plugin? It is webrtc related class. Am I wrong?
> @ibc reply: No, It shouldn't.

See https://github.com/cordova-rtc/cordova-plugin-iosrtc/issues/355#issuecomment-460240767

Skorpeo commented:
Hi, thanks for quick response.  

- Short answer: it will not work with adapter out of the box (and probably never will - out of the box).  

- Long Story: I would call it a hairy fix to make the plugin work with adapter, nonetheless the changes to this library are still worthwhile even if it doesn't work "out of the box" with adapter.  There are multiple issues that had to be addressed to make it work with adapter, adding addEventListener was only one problem, once I fixed that other issues arose that are also addressed in this PR (at least the ones I became aware of).  In short, it seems to me that adapter will never work with this library in its current form.  The main reason for this is that in a few places adapter calls `new MediaStream()` constructor, this library uses a factory function to instantiate a mediastream namely `MediaStream.create()` as MediaStream is actually a Blob().  I tried but it is not possible to subclass window.Blob(), i.e. there is no way we can create a class that conforms to MediaStream constructor while being a subclass of Blob().  At least in ES5; it may be possible in ES6 but I didn't check, and if we can use ES6 I would also suggest fixing the library to use classes instead of functions.  I created a new method for MediaStream.init() that for the most part mimicks MediaStream constructor with one notable difference that it requires a id.  So to make it work with adapter:

1. Any calls to `new MediaStream(`) have to be changed in adapter itself to call `MediaStream.init()` to make adapter work. 

2.  The build process has a nasty treatment of addEventListener that had my head scratching before I discovered the issue and it wasn't fun wasting 2 days trying to figure out why something that should work wasn't.  I won't get into the philosophical evaluation of my sanity during this time but to conclude the js file that the build process spits out specifically prevents any changes to addEventListener. To make it work I had to build and after it was built I had to comment out the portion that prevents adding the addEventListener method. 

So I guess if we can't include adapter file and assuming there is no way to have build process not include that nasty treatment I would suggest perhaps including a patch or script that will patch adapter and figure out something similar for build process, the latter being more stable?  My humble opinion is it would be nice to include in some sort of folder a working adapter implementation but I understand comments against it as well.  

At the end of the day I can make it work using above steps (at least for now).

Finally I understand the impurity and perversion of the above but my humble opinion is that I think this library has better chance of survival if it works and people can/do use it.  Like I mentioned before I am not a javascript/ios/cordova/webrtc guy but it seems to me that there's a lot of issues with webrtc and a lot of shims to make it work, case in point the adapter script and its many versions :/.  This library itself is a shim for ios.  So until the dust settles...... 

With that said I defer to you.

hthetiot commented:
Merged #373 fix confirmed, RTCPeerConnection.addTrack and webrtc-adapter now supported, will land in 5.0.2 release.

Thank you @Skorpeo make PR anytime. You really helped.

Skorpeo commented:
Np, glad it helped.  Here is a link to what the native constructor looks like in case it helps in the future: https://developer.mozilla.org/en-US/docs/Web/API/MediaStream/MediaStream
